pkg/codegen/importer/hcl2.go

```diff
@@ -0,0 +1,467 @@
+	})
+	return block
+}
+
+func makeResourceOptions(state *resource.State, names NameTable) (*model.Block, error) {
+	var resourceOptions *model.Block
+	if state.Parent != "" && state.Parent.Type() != resource.RootStackType {
+		name, ok := names[state.Parent]
+		if !ok {
+			return nil, fmt.Errorf("no name for parent %v", state.Parent)
+		}
+		resourceOptions = appendResourceOption(resourceOptions, "parent", newVariableReference(name))
+	}
+	if state.Provider != "" {
+		ref, err := providers.ParseReference(state.Provider)
+		if err != nil {
+			return nil, fmt.Errorf("invalid provider reference %v: %w", state.Provider, err)
+		}
+		if !providers.IsDefaultProvider(ref.URN()) {
+			name, ok := names[ref.URN()]
+			if !ok {
+				return nil, fmt.Errorf("no name for provider %v", state.Parent)
+			}
+			resourceOptions = appendResourceOption(resourceOptions, "provider", newVariableReference(name))
+		}
+	}
+	if len(state.Dependencies) != 0 {
+		deps := make([]model.Expression, len(state.Dependencies))
+		for i, d := range state.Dependencies {
+			name, ok := names[d]
+			if !ok {
+				return nil, fmt.Errorf("no name for resource %v", d)
+			}
+			deps[i] = newVariableReference(name)
+		}
+		resourceOptions = appendResourceOption(resourceOptions, "dependsOn", &model.TupleConsExpression{
+			Tokens:      syntax.NewTupleConsTokens(len(deps)),
+			Expressions: deps,
+		})
+	}
+	if state.Protect {
+		resourceOptions = appendResourceOption(resourceOptions, "protect", &model.LiteralValueExpression{
+			Tokens: syntax.NewLiteralValueTokens(cty.True),
+			Value:  cty.True,
+		})
+	}
+	return resourceOptions, nil
+}
+
+// typeRank orders types by their simplicity.
+func typeRank(t schema.Type) int {
+	switch t {
+	case schema.BoolType:
+		return 1
+	case schema.IntType:
+		return 2
+	case schema.NumberType:
+		return 3
+	case schema.StringType:
+		return 4
+	case schema.ArchiveType:
+		return 5
+	case schema.AssetType:
+		return 6
+	case schema.JSONType:
+		return 7
+	case schema.AnyType:
+		return 13
+	default:
+		switch t.(type) {
+		case *schema.TokenType:
+			return 8
+		case *schema.ArrayType:
+			return 9
+		case *schema.MapType:
+			return 10
+		case *schema.ObjectType:
+			return 11
+		case *schema.UnionType:
+			return 12
+		default:
+			return int(math.MaxInt32)
+		}
+	}
+}
+
+// simplerType returns true if T is simpler than U.
+//
+// The first-order ranking is:
+//
+//   bool < int < number < string < archive < asset < json < token < array < map < object < union < any
+//
+// Additional rules apply to composite types of the same kind:
+// - array(T) is simpler than array(U) if T is simpler than U
+// - map(T) is simpler than map(U) if T is simpler than U
+// - object({ ... }) is simpler than object({ ... }) if the former has a greater number of required properties that are
+//   simpler than the latter's required properties
+// - union(...) is simpler than union(...) if the former's simplest element type is simpler than the latter's simplest
```

Comment:
Is this preferable to ranking based on the most complex element type?

Reply:
Can you elaborate more on what you mean?

Reply:
Rather than saying simpler means "has the simplest type", you could also define simpler as "the most complex type is simpler than the other most complex type". Basically the difference between comparing lower bound and upper bound.

Reply:
I prefer defining this based on lower bounds. Picking the smallest lower bound guarantees that `zeroValue` will pick the simplest type out of the set of possible types, which should generate the simplest zero value.